### PR TITLE
feat(testing): introducing more tests and use cases coverage

### PR DESCRIPTION
**Pull request by tchataigner**
## Description

This PR is meant to cover more use cases that we previously missed on Kythera, while also testing that they properly work.

**Major changes**
- Call `Constructor` on target actors if it is relevant
- Fixed bug with sequencing by directly fetching correct sequence to use from `StateTree`
- `Tester::test` now tests only one test actor. This allows us to have better error handling and streamed results.
- Added more tests in CLI in an attempt to cover more cases that might happen to our users.

**Minor changes**
- Prettified a little bit the CLI output with indentation.
- Tried to sanitize and simplify testing context by removing some test actors.
- Changed some logs from `info` to `error` when relevant.
- Moved CLI tests to dedicated file.

## Open Questions

- Revamp CLI printing: Open #61 

## Change checklist

- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: lib/tests/lib.rs [hunk 1]
```diff
@@ -25,15 +17,39 @@ fn set_target_actor(tester: &mut Tester, name: String, binary: Vec<u8>, abi: Abi
...
+        },
+    );
+
+    let res = tester.deploy_target_actor(target_actor);
+    assert!(res
+        .err()
+        .as_ref()
+        .unwrap()
+        .to_string()
+        .contains("Non valid target actor wasm file: Target.wasm"))
```
Comment (jxs) [inline lib/tests/lib.rs:40]: I'd say that if we really want to be sure that this an error for this method we could create an `Error` variant for it

**Review by jxs — COMMENTED** [review]
very good work Thomas! Left some comments

File: lib/src/lib.rs [hunk 2]
```diff
@@ -203,159 +258,152 @@ impl Tester {
...
-                                }
-                                result
-                            })
-                            .collect(),
-                    ),
-                }
+                result
             })
-            .collect())
+            .collect::<Vec<TestResult>>())
```
[the remark this answers is not in the document]
Comment (tchataigner) [reply lib/src/lib.rs]: Yes ! I specified the type to ease the development of the map function !

File: cli/src/commands/test/mod.rs [hunk 3]
```diff
@@ -28,34 +31,67 @@ pub(crate) struct TestArgs {
...
-            .expect("Should be able to sync the end of streaming results");
+        if let Err(err) = tester.deploy_target_actor(test_target.actor) {
+            log::error!("\nError: {}", err);
+            if let Some(source) = err.source() {
+                log::error!("Caused by: {}", source)
+            }
+            continue;
+        };
+
+        let populated_tests = test_target
```
[the remark this answers is not in the document]
Comment (tchataigner) [reply cli/src/commands/test/mod.rs:48]: Oh good point, should have done that, thanks :pray:

File: cli/src/commands/test/mod.rs [hunk 4]
```diff
@@ -28,34 +31,67 @@ pub(crate) struct TestArgs {
...
+        log::info!("  Running Tests for Actor : {}", test_target.actor.name());
         let mut tester = Tester::new();
-        tester.deploy_target_actor(test.actor)?;
-        let verbosity = args.verbosity;
-        thread::spawn(move || stream_results(stream_rx, sync_tx, verbosity));
-        let _results = tester.test(&test.tests, Some(stream_tx))?;
-        sync_rx
-            .recv()
-            .expect("Should be able to sync the end of streaming results");
+        if let Err(err) = tester.deploy_target_actor(test_target.actor) {
```
[the remark this answers is not in the document]
Comment (tchataigner) [reply cli/src/commands/test/mod.rs:39]: Yes it makes sense !

[the remark this answers is not in the document]
Comment (tchataigner) [reply cli/src/commands/test/mod.rs:39]: Need to keep check on source as error can also be `Constructor` that might have no source (option type)

**Review by jxs — APPROVED** [review]

**Pull request merged**